server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -865,7 +865,9 @@ public Index(RecoveryFilesDetails recoveryFilesDetails) {
 
         public Index(StreamInput in) throws IOException {
             super(in);
-            fileDetails = new RecoveryFilesDetails(in);
+            synchronized (this) {
+                fileDetails = new RecoveryFilesDetails(in);
+            }
```

Comment:
I am not sure this is necessary. Here, `fileDetails` is a final field that is initialized here in `Index` constructor.

According to the https://docs.oracle.com/javase/specs/jls/se7/html/jls-17.html#jls-17.5
`An object is considered to be completely initialized when its constructor finishes. A thread that can only see a reference to an object after that object has been completely initialized is guaranteed to see the correctly initialized values for that object's final fields.`

My understanding is that (event though `fileDetails` itself is mutable) it will be completely initialized (read from `in` in this case) by the time another thread can have a chance to see it and lock for other updates or reads. Do I miss anything?

Reply:
You're right. I think we don't synchronized on this because `fileDetails` is final, so there's a read barrier that guarantees safe initialization